snooty/postprocess.py

```diff
@@ -1547,6 +1547,44 @@ def _attach_doc_title(self, fileid_stack: FileIdStack, node: n.RefRole) -> None:
         node.children = [deepcopy(node) for node in title]
 
 
+class FacetsHandler(Handler):
+    """Builds page.facets depending on facets found on nodes on this page"""
+
+    # TODO: should be able to handle facet propagation from project/directory/page -> page in the future
+    # when taxonomy and function of facets (propagation) are decided
+    def __init__(self, context: Context) -> None:
+        super().__init__(context)
+
+        self.facets: Dict[str, Union[List[object], str]] = {}
```

Comment:
Hm. I'd like to see concrete types here rather than `List[object]` if we can do that, whether by using TypedDict on the easier side of things:
```
class FacetDef(TypedDict, total=False):
    name: str
```

or a dataclass:
```
@dataclass
class FacetDef:
    name: str
    children: List[FacetDef]
```

Reply:
the taxonomy is TBD and i think this makes it hard for defining types beforehand, not knowing what attributes each level might hold. ie. `genres` might have subgenres(?) but `programming language` might have sub versions. ` would it be clearer to have the arbitrary fields under one `subfacets` field

Reply:
discussed offline. made usage of `SerializedNode` to prevent `object` typing here. it is a little bit more loose than intended, but the [sequence of serializableTypes](https://github.com/mongodb/snooty-parser/blob/master/snooty/n.py#L49) fits the need here. can be more explicit if need be.

given the first approach above, we would have to change the output / test assertions from:

```
facets == {
            "genres": [{"name": "reference"}, {"name": "tutorial"}],
            "target_platforms": [{"name": "atlas", "versions": [{"name": "v1.2"}]}],
        }
```

to convert to a list:
```
facets == [
                     { 'name': 'genres', 'children': [{"name": "reference"}, {"name": "tutorial"}]},
                     { 'name': 'target_platforms', 'children': [{"name": "atlas", "versions": [{"name": "v1.2"}]}],
        }]
```